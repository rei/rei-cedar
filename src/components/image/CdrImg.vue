<template>
  <div
    v-if="ratio"
    :style="ratioObject"
    :class="[style[ratioClass], containerClass]"
  >
    <img
      :style="cropObject"
      :class="mapClasses(style,
                         baseClass,
                         modifierClass,
                         radiusClass,
                         coverWrapperClass,
                         cropClass,
                         coverClass,
      )"
      :src="src"
      :alt="alt"
      v-bind="$attrs"
    >
  </div>
  <img
    v-else
    :class="mapClasses(style,
                       baseClass,
                       modifierClass,
                       radiusClass,
    )"
    :src="src"
    :alt="alt"
    v-bind="$attrs"
  >
</template>
<script>
// use normal <script> to declare options
export default {
  inheritAttrs: false
}
</script>
<script setup>
import { useCssModule, computed } from 'vue';
import mapClasses from '../../utils/mapClasses.js';
import propValidator from '../../utils/propValidator.js';
const props = defineProps({
  /**
   * Required. Image source url.
   */
  src: {
    type: String,
    required: true,
  },
  /**
   * Required. Image alt text.
   */
  alt: {
    type: String,
    default: '',
  },
  /**
   * Aspect ratio of the media container. {auto, square, 1-2, 2-3, 3-4, 9-16, 2-1, 3-2, 4-3, 16-9}
   */
  ratio: {
    type: String,
    validator: (value) => ([
      'auto',
      'square',
      '1-2',
      '2-3',
      '3-4',
      '9-16',
      '2-1',
      '3-2',
      '4-3',
      '16-9'].indexOf(value) >= 0) || false,
  },
  /**
   * Requires a `ratio`. Area to crop the image overflow to. {left, center, right} {top, center, bottom}
   */
  crop: {
    type: String,
  },
  /**
   * Requires a `ratio`. Scale the image to be as large as possible to fill the area (background-position: cover;)
   */
  cover: {
    type: Boolean,
  },
    /**
   * Adds a custom class to the cdr-img__ratio container div
   */
  containerClass: String,
  /**
   * Sets a border radius to an element {square, top, right, bottom, left}
   */
  radius: {
    type: String,
    validator: (value) => ([
      'circle',
      'rounded'].indexOf(value) >= 0) || false,
  },

  modifier: {
    type: String,
    default: '',
    validator: (value) => propValidator(value, ['', 'responsive']),
  }  
});
const style = useCssModule();
const baseClass = 'cdr-image';
const ratioClass = 'cdr-image-ratio';
const coverWrapperClass = 'cdr-image-ratio__cover';
const modifierClass = computed(() => props.modifer && `${baseClass}--${props.modifer}`);
const radiusClass = computed(() => props.radius && `${baseClass}--${props.radius}`);
const cropObject = computed(() => ({ objectPosition: props.crop }));
const ratioObject = computed(() => {
  let ratioPct;
  if (props.ratio === 'square') {
    ratioPct = '100%';
  } else if (props.ratio === 'auto') {
    ratioPct = '0';
  } else {
    const [x, y] = props.ratio.split('-');
    ratioPct = `${(y / x) * 100}%`;
  }
  return { '--ratio': ratioPct };
});

const cropClass = computed(() => props.crop && `${coverWrapperClass}--crop`);
const coverClass = computed(() => props.cover && `${coverWrapperClass}--cover`);

</script>

<style lang="scss" module src="./styles/CdrImg.module.scss">
</style>
