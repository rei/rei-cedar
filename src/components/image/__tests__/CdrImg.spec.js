import { shallowMount, mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrImg from '../CdrImg.vue';
import sinon from 'sinon';

describe('CdrImg', () => {
  describe('with just the alt text set', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrImg, {
        propsData: {
          src: '/src/dev/static/cedar-350x150.jpg',
          alt: 'test alt',
        }
      });
    })

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('sets the alt attr correctly', () => {
      expect(wrapper.attributes().alt).toBe('test alt');
    });
  })

  describe('when crop and ratio and radius and responsive is set', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrImg, {
        propsData: {
          ratio: "square",
          radius: "rounded",
          modifier: "responsive",
          cover: true,
          crop: "left",
          alt: "crop left",
          src: "/src/dev/static/cedar-1920x1080.jpg",
        }
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('a media frame', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallowMount(CdrImg, {
        propsData: {
          src: '/src/dev/static/cedar-350x150.jpg',
          ratio: 'square',
          alt: 'test alt',
        }
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('sets the alt attr correctly', () => {
      expect(wrapper.find('img').attributes().alt).toBe('test alt');
    });
  });

  describe('when arbitrary HTML attrs are passed', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrImg, {
        propsData: {
          src: '/src/dev/static/cedar-350x150.jpg',
          loading: 'lazy',
          ratio: 'square'
        }
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('the arbitrary attrs are passed to the img element', () => {
      expect(wrapper.find('img').attributes()['loading']).toBe('lazy');
    });
  });

  describe('', () => {
    let wrapper;
    let spy;
    beforeEach(() => {
      spy = sinon.spy();
      wrapper = shallowMount(CdrImg, {
        propsData: {
          src: 'localhost:8000/nothing-to-see-here.png',
          onError: spy,
        }
      });
      wrapper.find('img').trigger('error');
    });

    it('emits error event for default image', () => {
      expect(spy.calledOnce).toBeTruthy();
    })
  });

  describe('image with auto ratio', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallowMount(CdrImg, {
        propsData: {
          src: 'localhost:8000/nothing-to-see-here.png',
          ratio: 'auto',
        },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('has the expected ratio object value', () => {
      expect(wrapper.vm.ratioObject['--ratio']).toBe('0');
    });
  });


  describe('image with 3-2 ratio', () => {
    let wrapper;
    let spy;
    beforeEach(() => {
      spy = sinon.spy();
      wrapper = shallowMount(CdrImg, {
        propsData: {
          src: 'localhost:8000/nothing-to-see-here.png',
          ratio: '3-2',
          onError: spy
        },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('has the expected ratio object value', () => {
      expect(wrapper.vm.ratioObject['--ratio']).toBe('66.66666666666666%');
    });

    it('emits error event for ratio image', () => {
      wrapper.find('img').trigger('error');
      expect(spy.calledOnce).toBeTruthy();
    })
  });
});
