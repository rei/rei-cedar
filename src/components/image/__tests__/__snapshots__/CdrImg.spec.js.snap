// Vitest Snapshot v1

exports[`CdrImg > renders correctly 1`] = `
<img
  alt="test alt"
  class="cdr-image"
  src="/src/dev/static/cedar-350x150.jpg"
/>
`;

exports[`CdrImg > renders crop/ratio correctly 1`] = `
<div
  class="cdr-image-ratio"
  style="--ratio: 100%;"
>
  <img
    alt="crop left"
    class="cdr-image cdr-image-ratio__cover cdr-image-ratio__cover--crop cdr-image-ratio__cover--cover"
    src="/src/dev/static/cedar-1920x1080.jpg"
    style="object-position: left;"
  />
</div>
`;
